app/services/csv_exporter.rb

```diff
@@ -1,25 +1,59 @@
 require 'csv'
 
 module CsvExporter
-  def self.export(resources, columns, header = nil)
-    header ||= default_header(columns)
-    raise ArgumentError, "Columns and header row aren't the same length" unless columns.length == header.length
-    # need to save the current context as the enumerator is executed by a separate thread
-    context = Foreman::ThreadSession::Context.get
-
-    Enumerator.new do |csv|
-      Foreman::ThreadSession::Context.set(**context)
-      csv << CSV.generate_line(header)
-      cols = columns.map { |c| c.to_s.split('.').map(&:to_sym) }
-      resources.uncached do
-        resources.reorder(nil).limit(nil).find_each do |obj|
-          csv << CSV.generate_line(cols.map { |c| c.inject(obj, :try) })
-        end
+  class ExportDefinition
+    attr_reader :label
+
+    def initialize(key, label: nil, callback: nil)
+      @key = key.to_s.split('.').map(&:to_sym)
+      @label = label || derive_label(key)
+      @callback = callback
+    end
+
+    def derive_label(key)
+      key.to_s.titleize.gsub('.', ' - ')
+    end
+
+    def evaluate(object)
+      if @callback
+        @callback.call(object)
+      else
+        @key.inject(object, :try)
       end
     end
   end
 
-  def self.default_header(columns)
-    columns.map { |c| c.to_s.titleize }
+  class << self
+    def export(resources, columns, header = nil)
+      columns = preprocess_columns(columns)
+      header ||= default_header(columns)
+      raise ArgumentError, "Columns and header row aren't the same length" unless columns.length == header.length
+      # need to save the current context as the enumerator is executed by a separate thread
+      context = Foreman::ThreadSession::Context.get
+
+      Enumerator.new do |csv|
+        Foreman::ThreadSession::Context.set(**context)
+        csv << CSV.generate_line(header)
+        resources.uncached do
+          resources.reorder(nil).limit(nil).find_each do |obj|
+            csv << CSV.generate_line(columns.map { |c| c.evaluate(obj) }.flatten)
+          end
+        end
+      end
+    end
+
+    def preprocess_columns(columns)
```

Comment:
A nitpick, that I don't mind if it will not be addressed:
we can skip this method, if we do the preprocessing in `csv_columns`. I mean making sure the method returns an array of `ExportDefinition` structures. We can use a factory method on `ExportDefinition` class to encapsulate the `if` statement.